src/firecracker/examples/uffd/uffd_utils.rs

```diff
@@ -263,13 +330,121 @@ impl UffdHandler {
     }
 }
 
+struct UffdMsgIterator {
+    stream: UnixStream,
+    buffer: Vec<u8>,
+    fds_received: bool,
+    secret_free: bool,
+    current_pos: usize,
+    fds: [RawFd; 3],
+}
+
+impl Iterator for UffdMsgIterator {
+    type Item = UffdMsgFromFirecracker;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if !self.fds_received {
+            let mut iov = [];
+            let mut fds = [0; 3];
+            let ret = unsafe { self.stream.recv_with_fds(&mut iov, &mut fds) };
+            match ret {
+                Ok((0, 1)) => {
+                    self.secret_free = false;
+                }
+                Ok((0, 3)) => {
+                    self.secret_free = true;
+                }
+                Ok((n, _)) => {
+                    panic!("Wrong number of fds: {}", n);
+                }
+                Err(e) => {
+                    if e.errno() == libc::EAGAIN {
+                        return None;
+                    }
+                    panic!("Read error: {}", e);
+                }
+            }
+            self.fds_received = true;
+            self.fds = fds;
+
+            match self.recv_mappings() {
+                Some(mappings) => Some(UffdMsgFromFirecracker::Mappings(mappings)),
+                None => None, // EOF or error
+            }
+        } else {
+            match self.recv_fault_request() {
+                Some(request) => Some(UffdMsgFromFirecracker::FaultReq(request)),
+                None => None, // EOF or error
+            }
+        }
+    }
+}
+
+impl UffdMsgIterator {
+    fn new(stream: UnixStream) -> Self {
+        Self {
+            stream,
+            buffer: vec![0u8; 4096],
+            fds_received: false,
+            secret_free: false,
+            current_pos: 0,
+            fds: [0; 3],
+        }
+    }
+
+    fn recv_json<T: serde::de::DeserializeOwned + serde::Serialize>(&mut self) -> Option<T> {
```

Comment:
Think you can drop `+serde::Serialize` here? compiled without it for me at least